vhdl_syntax/src/parser/productions/compositve_types.rs

```diff
@@ -0,0 +1,252 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c)  2025, Lukas Scheller lukasscheller@icloud.com
+/// Parsing of composite types (LRM §5.3)
+use crate::parser::Parser;
+use crate::syntax::node_kind::NodeKind::*;
+use crate::tokens::token_kind::Keyword as Kw;
+use crate::tokens::TokenKind::*;
+use crate::tokens::TokenStream;
+
+impl<T: TokenStream> Parser<T> {
+    pub fn array_type_definition(&mut self) {
+        self.start_node(ArrayTypeDefinition);
+        self.expect_kw(Kw::Array);
+        let box_found = self.lookahead_skip_n(1, [BOX]).is_ok();
+
+        if box_found {
+            self.expect_token(LeftPar);
+            self.separated_list(Parser::index_subtype_definition, Comma);
+            self.expect_token(RightPar);
+        } else {
+            self.index_constraint();
+        }
+        self.expect_kw(Kw::Of);
+        self.subtype_indication();
+        self.end_node();
+    }
+
+    pub fn record_type_definition(&mut self) {
+        self.start_node(RecordTypeDefinition);
+        self.expect_kw(Kw::Record);
+
+        while !self.next_is(Keyword(Kw::End)) {
+            self.element_declaration();
+        }
+
+        self.expect_kw(Kw::End);
+        self.expect_kw(Kw::Record);
+        self.opt_identifier();
+        self.end_node();
+    }
+
+    pub fn element_declaration(&mut self) {
+        self.start_node(ElementDeclaration);
+        self.identifier_list();
+        self.expect_token(Colon);
+        self.subtype_indication();
+        self.expect_token(SemiColon);
+        self.end_node();
+    }
+
+    pub fn index_subtype_definition(&mut self) {
+        self.start_node(IndexSubtypeDefinition);
+        self.type_mark();
+        self.expect_tokens([Keyword(Kw::Range), BOX]);
+        self.end_node();
+    }
+
+    pub fn index_constraint(&mut self) {
+        self.start_node(IndexConstraint);
+        self.expect_token(LeftPar);
+        self.separated_list(Parser::discrete_range, Comma);
+        self.expect_token(RightPar);
+        self.end_node();
+    }
+
+    pub fn discrete_range(&mut self) {
+        todo!();
```

Comment:
We could work around `discrete_range` using the `RawTokens` node. So constrained array definitions would be (partially) parseable.

Reply:
That sounds good to me 👍